presto-main/src/main/java/com/facebook/presto/operator/RemoteProjectOperator.java

```diff
@@ -0,0 +1,204 @@
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class RemoteProjectOperator
+        implements Operator
+{
+    private final OperatorContext operatorContext;
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final List<RowExpression> projections;
+
+    private final CompletableFuture<SqlFunctionResult>[] result;
+
+    private boolean finishing;
+
+    private RemoteProjectOperator(OperatorContext operatorContext, FunctionAndTypeManager functionAndTypeManager, List<RowExpression> projections)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionManager is null");
+        this.projections = ImmutableList.copyOf(requireNonNull(projections, "projections is null"));
+        this.result = new CompletableFuture[projections.size()];
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && !processingPage();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        checkState(!processingPage(), "Still processing previous input");
+        requireNonNull(page, "page is null");
+        for (int channel = 0; channel < projections.size(); channel++) {
+            RowExpression projection = projections.get(channel);
+            if (projection instanceof InputReferenceExpression) {
+                result[channel] = completedFuture(new SqlFunctionResult(page.getBlock(((InputReferenceExpression) projection).getField()), 0));
+            }
+            else if (projection instanceof CallExpression) {
+                CallExpression remoteCall = (CallExpression) projection;
+                result[channel] = functionAndTypeManager.executeFunction(
+                        operatorContext.getDriverContext().getTaskId().toString(),
+                        remoteCall.getFunctionHandle(),
+                        page,
+                        remoteCall.getArguments().stream()
+                                .map(InputReferenceExpression.class::cast)
+                                .map(InputReferenceExpression::getField)
+                                .collect(toImmutableList()));
+            }
+            else {
+                checkState(projection instanceof ConstantExpression, format("Does not expect expression type %s", projection.getClass()));
+            }
+        }
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (resultReady()) {
+            Block[] blocks = new Block[result.length];
+            Page output;
+            try {
+                for (int i = 0; i < blocks.length; i++) {
+                    blocks[i] = result[i].get().getResult();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It looks like you are using CompletableFuture.get() without setting its timeout parameter. Be sure to set timeout values or handle timeout exceptions to prevent threads from waiting for a process to end. This can lead to downtime for your customers. [here](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#get-long-java.util.concurrent.TimeUnit-).